draft-ietf-ppm-dap.md

```diff
@@ -1214,64 +1269,174 @@ as follows:
+`AggregationJob.round` MUST be set to the Leader's current round (i.e., one more
+than the `round` value in the Helper's latest `AggregationJob` message) and the
+`prepare_steps` field MUST be a sequence of `PrepareStep`s in the `continued`
+state containing the corresponding `inbound` prepare message.
+
+###### Helper Continuation
+
+If the Helper does not recognize the task ID, then it MUST abort with error
+`unrecognizedTask`. If the Helper does not recognize the aggregation job, then
+it MUST abort with error `unrecognizedAggregationJob`. This could occur because
+the aggregation job never existed, but also if it was deleted (see
+{{aggregation-job-delete}}).
+
+Otherwise, for each `PrepareStep` received from the Leader in
+`AggregationJob.prepare_steps`, the Helper checks
+`PrepareStep.prepare_step_state` to determine if the report share should
+continue being prepared. If the report share is in state `failed` or `finished`,
+then the Helper marks the report as failed with error `unrecognized_message`.
+
+If the state is `continued`, then the Helper continues with preparation for the
+report share. The Helper MUST check its current preparation state against the
+Leader's `AggregationJob.round` value. If the Leader is one round ahead of the
+Helper, then the Helper can combine the Leader's prepare message and the
+Helper's current preparation state (`prep_state`) as follows:
 
-struct {
-  PrepareStep prepare_steps<1..2^32-1>;
-} AggregateInitializeResp;
 ~~~
+out = VDAF.prep_next(prep_state, inbound)
+~~~
+
+where `inbound` is the previous VDAF prepare message sent by the Leader and
+`prep_state` is the Helper's current preparation state. This step yields one of
+three outputs:
+
+1. An error, in which case the input report share is marked as failed and the
+   Helper replies to the Leader with a `PrepareStep` in the `failed` state.
+1. An output share, in which case the Helper stores the output share for future
+   collection as described in {{collect-flow}} and replies to the Leader with a
+   `PrepareStep` in the `finished` state. The Helper expects no further requests
+   from the Leader for a report in the `finished` state.
+1. An updated VDAF state and preparation message, denoted
+   `(prep_state, prep_msg)`, in which case the Helper replies to the leader with
+   `PrepareStep` in the `continued` state containing `prep_msg`.
 
-The message is a sequence of PrepareStep values, the order of which matches that
-of the ReportShare values in `AggregateInitializeReq.report_shares`. Each report
-that was marked as invalid is assigned the PrepareStepResult `failed`.
-Otherwise, the PrepareStep is either marked as continued with the output
-`prep_msg`, or is marked as finished if the VDAF preparation process is finished
-for the report share.
+After computing each report share's preparation state and prepare message, the
+Helper SHOULD store them, tagged with the round they were computed in, so that
```

Comment:
Here we say the helper should store the answer, but below we say the helper can either store or recompute the response (paragraph starting L1404).

I'd suggest dropping the "recompute" part -- storing the answer is simpler & less error-prone. And recomputing the response still requires storing information to allow the recomputation to occur, I think.

Reply:
I think this is similar to the equivalent leader text discussed [here](https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/pull/395#discussion_r1067581728) so I'll pursue discussion there.

Reply:
The only part of this thread that's different from the linked thread is that this thread suggests dropping the suggestion to "recompute" -- IIUC, recomputing is a strictly inferior strategy to storing the previously-computed result (since even in a recompute strategy, you still need to store the data required to recompute the result, i.e. recomputation also requires storage).

Agreed w/ your comment in the other thread that we should probably not be very specific about when storage occurs.

Reply:
I deleted the SHOULDs about storing anything and punted the discussion of round skew to a new "Recovering From Round Skew".